language/move-model/src/builder/bytecode_builder.rs

```diff
@@ -0,0 +1,171 @@
+                type_params,
+                None, // Some(fields)
+            );
+        }
+
+        for f in normalized.public_functions {
+            let is_public = true;
+            make_fun(&mut builder, module_name.clone(), module_id, f, is_public)
+        }
+        for f in private_functions {
+            let is_public = false;
+            make_fun(&mut builder, module_name.clone(), module_id, f, is_public)
+        }
+
+        let mut module_translator = ModuleBuilder::new(&mut builder, module_id, module_name);
+        module_translator.populate_env_from_result(Loc::default(), m, SourceMap::new(None));
+    }
+    Ok(env)
+}
+
+fn make_type_params(builder: &mut ModelBuilder, kinds: &[Kind]) -> Vec<(Symbol, Type)> {
+    kinds
+        .iter()
+        .enumerate()
+        .map(|(i, _)| {
+            let dummy_name = builder.env.symbol_pool().make(&i.to_string());
+            (dummy_name, Type::TypeParameter(i as u16))
+        })
+        .collect()
+}
+
+fn make_parameters(
+    builder: &mut ModelBuilder,
+    types: Vec<normalized::Type>,
+) -> Vec<(Symbol, Type)> {
+    types
+        .into_iter()
+        .enumerate()
+        .map(|(i, t)| {
+            let dummy_name = builder.env.symbol_pool().make(&i.to_string());
+            (dummy_name, make_type(builder, t))
+        })
+        .collect()
+}
+
+fn make_type(builder: &mut ModelBuilder, t: normalized::Type) -> Type {
+    use PrimitiveType::*;
+    use Type::*;
+    match t {
+        normalized::Type::Bool => Primitive(Bool),
+        normalized::Type::U8 => Primitive(U8),
+        normalized::Type::U64 => Primitive(U64),
+        normalized::Type::U128 => Primitive(U128),
+        normalized::Type::Address => Primitive(Address),
+        normalized::Type::Signer => Primitive(Signer),
+        normalized::Type::TypeParameter(i) => TypeParameter(i),
+        normalized::Type::Reference(t) => Reference(false, Box::new(make_type(builder, *t))),
+        normalized::Type::MutableReference(t) => Reference(true, Box::new(make_type(builder, *t))),
+        normalized::Type::Vector(t) => Vector(Box::new(make_type(builder, *t))),
+        normalized::Type::Struct {
+            address,
+            module,
+            name,
+            type_arguments,
+        } => {
+            let addr = ModuleEnv::addr_to_big_uint(&address);
+            let module_name =
+                ModuleName::new(addr, builder.env.symbol_pool().make(module.as_str()));
+            let types = type_arguments
+                .into_iter()
+                .map(|t| make_type(builder, t))
+                .collect();
+            let menv = builder.env.find_module(&module_name).unwrap_or_else(|| {
+                panic!(
+                "Couldn't resolve module {:?}; make sure input module list is sorted in dep order",
+                module
+            )
+            });
+            let struct_id = menv
+                .find_struct_by_identifier(name.clone())
```

Comment:
Alternatively to doing two phases or topsort, you can construct a StructId from scratch without needing to have it defined already in the environment, because it is defined as StructId(Symbol), and you can assume well-formedness of bytecode.